lib/ddtrace/workers.rb

```diff
@@ -106,12 +107,28 @@ def perform
         loop do
           @back_off = flush_data ? @flush_interval : [@back_off * BACK_OFF_RATIO, BACK_OFF_MAX].min
 
+          record_metrics
+
           @mutex.synchronize do
             return if !@run && @trace_buffer.empty?
             @shutdown.wait(@mutex, @back_off) if @run # do not wait when shutting down
           end
         end
       end
+
+      def record_metrics
+        if Utils::Time::THREAD_CPU_TIME_SUPPORTED
+          Datadog.health_metrics.writer_cpu_time { thread_cpu_time_diff }
+        end
+      end
+
+      # Difference in seconds of CPU time since last measurement
+      def thread_cpu_time_diff
```

Comment:
I don't think this measurement behavior should be in here; maybe `Datadog::Utils::Time` would be more appropriate?

FWIW I added something just like this in https://github.com/DataDog/dd-trace-rb/pull/1017/commits/9d16767b2153be94bd113afc66fa7fd2da12b565#diff-0d88a184523b0c51a78d8ac5964d92f0R13 although its pending in the `feature/profiling` branch. It might be good to combine functionality here, and have a function like `measure` that takes a clock type, just like https://ruby-doc.org/core-2.1.0/Process.html#method-c-clock_gettime does.

Reply:
Hey David, I think we are talking about two distinct types of measurement here.

What I want to measure here is a continuous metric, with no gaps, starting at a specific point in time (thread start time in this case). We are pretty much plotting a graph for the cumulative total CPU time since thread inception, but capturing the value provided by the system: total cumulative thread CPU time.
Because the value provided to us is cumulative, we plot it as delta intervals: the integral of this plot at time _t_ will equal the system provided cumulative value at _t_.

This is in contrast with measuring a block (like `measure { ... }`), which will capture the time spent inside that block. Adding up measurements from `measure { ... }` will yield the total time spent inside that block, which is an interesting measurement to consider. With this in mind, this `measure { ... }` is not the same as plotting the continuous total CPU time tracked by the system: they likely won't be too far off, but `measure { ... }` is discrete, while capturing total CPU in deltas is continuous. There will be gaps in `measure { ... }`, while total CPU time will be a continuous measurement.

So with that in mind, we have to capture total CPU time using deltas, which require us to store the previous value to compare with the next measurement. This can be extracted to a separate stateful object, if that's a concern:

```ruby
def initialize
  @tracker = Datadog::Util::TimeTracker.new
end

def perform
  Datadog.health_metrics.writer_cpu_time { @tracker.get_thread_cpu_time_delta }
  # ...
end
```

The reasons I didn't are:
* We have two places using this code today (both current and new worker classes) and one of them is replacing the other, so we'll only have one place using it.
* The abstraction is somewhat leaky: the tracker object is still being called on the same call sites as the the old one. This can be argued, it just didn't cross my personal threshold of warranting a context switch to understand what's happening vs having the logic here.

We can still have `measure { ... }` in the worker thread, I think it will be very much valuable, specially to measure specific parts (serialization vs pipeline vs transport overhead).

While thread cpu time is a high level metric, `measure { ... }` is a granular metric: both provide their own value, but are two distinct kinds of measurement.

The way I see it, it is analogous to:
* Measuring high level application CPU time => `Utils::Time.get_thread_cpu_time`
* Tracing application-specific components with `ddtrace` => `measure { ... }`